### PR TITLE
Zc/single cell linkers

### PR DESCRIPTION
**Pull request by corps**
Generalized single cell linker functions, with override support!

Now setting up single cell linking is just a function call away!

eg:

```
@metis_etl(...)
def my_etl(helpers):      
    helpers.link_single_cell_attribute_files_v1(
        'sc_rna_seq',
        'MVIR1-HS',
    )
    
    helpers.link_single_cell_attribute_files_v1(
        'sc_rna_seq_pool',
        'MVIR1-POOL',
    )
```

### CODE REVIEW

File: airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+from airflow.utils.task_group import TaskGroup
+from etna.etls.metis import MatchedRecordFolder
+
+from etna import MetisEtlHelpers, link
+
+timepoint_regex = re.compile(r"^(?P<timepoint>[A-Z0-9]+-[A-Z0-9]+-[A-Z][A-Z]?\d+)[A-Z0-9]+.*")
+patient_regex = re.compile(r"^(?P<patient>[A-Z0-9]+-[A-Z0-9]+)-.*")
+day_regex = re.compile(r".*-[A-Z][A-Z]?(?P<day>\d+)$")
+
+# TODO:  Would be great if a naming service provided this structure for us.
```
Comment (corps) [inline airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py:23]: This is the start of the meat of it.

File: airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import re
+
+from airflow.exceptions import AirflowException
+from airflow.models.xcom_arg import XComArg
+from airflow.utils.task_group import TaskGroup
+from etna.etls.metis import MatchedRecordFolder
+
+from etna import MetisEtlHelpers, link
+
+timepoint_regex = re.compile(r"^(?P<timepoint>[A-Z0-9]+-[A-Z0-9]+-[A-Z][A-Z]?\d+)[A-Z0-9]+.*")
```
Comment (corps) [inline airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py:19]: @arkal 
Generalized these regexes so they can attempt to process even specialized identifiers.

Note below that I throw and fail all processing though if a record folder does not even match these wide regexes, as that would generally imply incorrect naming of something.

File: airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    with TaskGroup(f"{model_name}_raw"):
+        raw_matches = helpers.find_record_folders(
+            model_name,
+            re.compile(f"^{single_cell_root_prefix}[^/]*/raw/{identifier_prefix}[^/]+"),
+        )
+
+        raw_matches = link(dry_run=dry_run)(create_and_link_parents)(raw_matches)
+        # create new records if needed
+        listed_matches = helpers.list_match_folders(raw_matches)
+        # make this a raw_fastq_files collection
```
Comment (corps) [inline airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py:104]: removing all comment, migrated to file_collection already for comet.

**Comment by graft** [comment]
😢 This is beautiful; a moment years in the waiting!

File: airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+from airflow.exceptions import AirflowException
+from airflow.models.xcom_arg import XComArg
+from airflow.utils.task_group import TaskGroup
+from etna.etls.metis import MatchedRecordFolder
+
+from etna import MetisEtlHelpers, link
+
+timepoint_regex = re.compile(r"^(?P<timepoint>[A-Z0-9]+-[A-Z0-9]+-[A-Z][A-Z]?\d+)[A-Z0-9]+.*")
+patient_regex = re.compile(r"^(?P<patient>[A-Z0-9]+-[A-Z0-9]+)-.*")
+day_regex = re.compile(r".*-[A-Z][A-Z]?(?P<day>\d+)$")
```
Comment (coleshaw) [inline airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py:21]: Will this account for "negative" days, like `DN1` in MVIR1?

**Review by coleshaw — APPROVED** [review]

File: airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+# TODO:  Would be great if a naming service provided this structure for us.
+def create_and_link_parents(matches: List[MatchedRecordFolder]):
+    for match in matches:
+        if match.model_name == "sc_rna_seq_pool":
```
Comment (coleshaw) [inline airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py:26]: Hm, will it be a problem that `coprojects_template` and I assume projects based off of it, use the model names `sc_seq` and `sc_seq_pool` without the `_rna` part?

Comment (corps) [reply airflow/opt/providers/etna/etna/etls/linkers/single_cell_rna_seq.py:26]: Good, point.  I can change this logic to look for `pool` or not `pool` to keep it generic.

**Pull request merged**